### PR TITLE
feat(rpc/ots): implement ots_getTransactionBySenderAndNonce

### LINKED ISSUE #8905: Implement ots_getTransactionBySenderAndNonce
### Describe the feature

https://github.com/paradigmxyz/reth/blob/2219d7d23e3b68118652ad5764005b56bb0dd1f3/crates/rpc/rpc-api/src/otterscan.rs#L77-L83

this should be very similar to #8638

and would likely require some additional provider trait functions that give access to the indexes + tracing

but we can start this by checking the pool:

https://github.com/paradigmxyz/reth/blob/2219d7d23e3b68118652ad5764005b56bb0dd1f3/crates/transaction-pool/src/traits.rs#L335-L340

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jsvisa**
implement and close #8905

### CODE REVIEW

File: crates/rpc/rpc/src/otterscan.rs [hunk 1]
```diff
@@ -198,10 +198,61 @@ where
...
+        sender: Address,
+        nonce: u64,
+    ) -> RpcResult<Option<TxHash>> {
+        // Check if the sender is a contract
+        if self.has_code(sender, None).await? {
+            return Ok(None)
+        }
+
+        let highest =
+            EthApiServer::transaction_count(&self.eth, sender, None).await?.saturating_to::<u64>();
```
Comment (emhane) [inline crates/rpc/rpc/src/otterscan.rs:210]: the method name `transaction_count` isn't used in `OtterscanServer`, so no need to call it explicitly with trait like this
```suggestion
            self.eth.transaction_count(sender, None).await?.saturating_to::<u64>();
```

File: crates/rpc/rpc/src/otterscan.rs [hunk 2]
```diff
@@ -198,10 +198,61 @@ where
...
+
+        // If the nonce is higher than the highest nonce, the transaction is pending or not exists
+        if nonce >= highest {
```
Comment (emhane) [inline crates/rpc/rpc/src/otterscan.rs:213]:
```suggestion
        if nonce > highest {
```

**Review by emhane — CHANGES_REQUESTED** [review]

File: crates/rpc/rpc/src/otterscan.rs (hunk 2 above)
Comment (emhane) [reply crates/rpc/rpc/src/otterscan.rs:213]: or is it code comment that needs update?

Comment (jsvisa) [reply crates/rpc/rpc/src/otterscan.rs:213]: thanks, update the comments

File: crates/rpc/rpc/src/otterscan.rs [hunk 3]
```diff
@@ -198,10 +198,62 @@ where
...
+        let mut high = self.eth.block_number()?.saturating_to::<u64>();
+        let mut num = high;
+
+        while low <= high {
+            let mid = (low + high) / 2;
+            let mid_nonce = EthApiServer::transaction_count(
+                &self.eth,
+                sender,
+                Some(BlockId::Number(BlockNumberOrTag::Number(mid))),
+            )
```
Comment (emhane) [inline crates/rpc/rpc/src/otterscan.rs:231]:
```suggestion
            let mid_nonce = self.eth.transaction_count(
                sender,
                Some(BlockId::Number(BlockNumberOrTag::Number(mid))),
            )
```

File: crates/rpc/rpc/src/otterscan.rs [hunk 4]
```diff
@@ -198,10 +198,62 @@ where
...
+            .await?
+            .saturating_to::<u64>();
+
+            // The `transaction_count` returns the `nonce` after the transaction was executed, which
+            // is the state of the account after the block, and we need to find the
+            // transaction whose nonce is the pre-state, so need to compare with
+            // `nonce+1`.
+            if mid_nonce < nonce + 1 {
```
Comment (emhane) [inline crates/rpc/rpc/src/otterscan.rs:239]:
```suggestion
            if mid_nonce == nonce + 1 {
                num = mid;
                break
            } else if mid_nonce <= nonce {
```

File: crates/rpc/rpc/src/otterscan.rs [hunk 5]
```diff
@@ -198,10 +198,62 @@ where
...
+                low = mid + 1; // not found in current block, need to search in the later blocks
+            } else {
+                high = mid - 1; // found in current block, try to find a lower block
+                num = mid;
```
Comment (emhane) [inline crates/rpc/rpc/src/otterscan.rs:243]:
```suggestion
                high = mid - 1; // not found in current block, try to find a lower block
```

File: crates/rpc/rpc/src/otterscan.rs [hunk 6]
```diff
@@ -198,10 +198,62 @@ where
...
+            }
+        }
+
+        let Some(BlockTransactions::Full(transactions)) =
+            self.eth.block_by_number(num.into(), true).await?.map(|block| block.inner.transactions)
```
Comment (emhane) [inline crates/rpc/rpc/src/otterscan.rs:248]:
```suggestion
            self.eth.block_by_number((num - 1).into(), true).await?.map(|block| block.inner.transactions)
```

**Review by emhane — DISMISSED** [review]
left to-do:

- move the binary search algorithm out to a new method
- tests for the binary search method
- before checking mid block, check high block and return early if the high block has nonce zero

- preferably some heuristics for optimisation: if high block has tx nonce 1 - 5, start with mid set to ¾ high.

**Comment by jsvisa** [comment]
@emhane seems we cant call the `transaction_count` directly, see the compile error as below:

```
error[E0034]: multiple applicable items in scope
   --> crates/rpc/rpc/src/otterscan.rs:227:38
    |
227 | ...self.eth.transaction_count(
    |             ^^^^^^^^^^^^^^^^^ multiple `transaction_count` found
    |
note: candidate #1 is defined in the trait `EthApiServer`
   --> /md0/code/reth/crates/rpc/rpc-eth-api/src/core.rs:169:5
    |
169 | /     async fn transaction_count(
170 | |         &self,
171 | |         address: Address,
172 | |         block_number: Option<BlockId>,
173 | |     ) -> RpcResult<U256>;
    | |_________________________^
note: candidate #2 is defined in the trait `LoadState`
   --> /md0/code/reth/crates/rpc/rpc-eth-api/src/helpers/state.rs:227:5
    |
227 | /     fn transaction_count(
228 | |         &self,
229 | |         address: Address,
230 | |         block_id: Option<BlockId>,
231 | |     ) -> impl Future<Output = E...
232 | |     where
233 | |         Self: SpawnBlocking,
    | |____________________________^
help: disambiguate the method for candidate #1
    |
227 |             let mid_nonce = EthApiServer::transaction_count(&self.eth, sender, Some(BlockId::Number(BlockNumberOrTag::Number(mid))))
    |                             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
help: disambiguate the method for candidate #2
    |
227 |             let mid_nonce = LoadState::transaction_count(&self.eth, sender, Some(BlockId::Number(BlockNumberOrTag::Number(mid))))
```

> move the binary search algorithm out to a new method
> tests for the binary search method

I'll try to move out and add tests.


> before checking mid block, check high block and return early if the high block has nonce zero

not fully understand, we have already check the highest block before the binary search.


> preferably some heuristics for optimisation: if high block has tx nonce 1 - 5, start with mid set to ¾ high.

in this pr, I'll try to keep it is as simple, and maybe optimize the binary search later(with some data analyze)

File: crates/rpc/rpc/src/otterscan.rs (hunk 6 above)
Comment (jsvisa) [reply crates/rpc/rpc/src/otterscan.rs:247]: I think the change is not correct, if found nonce in block N, need to get txs in block N, not N-1

**Comment by emhane** [comment]
> @emhane seems we cant call the `transaction_count` directly, see the compile error as below:
> 
> ```
> error[E0034]: multiple applicable items in scope
>    --> crates/rpc/rpc/src/otterscan.rs:227:38
>     |
> 227 | ...self.eth.transaction_count(
>     |             ^^^^^^^^^^^^^^^^^ multiple `transaction_count` found
>     |
> note: candidate #1 is defined in the trait `EthApiServer`
>    --> /md0/code/reth/crates/rpc/rpc-eth-api/src/core.rs:169:5
>     |
> 169 | /     async fn transaction_count(
> 170 | |         &self,
> 171 | |         address: Address,
> 172 | |         block_number: Option<BlockId>,
> 173 | |     ) -> RpcResult<U256>;
>     | |_________________________^
> note: candidate #2 is defined in the trait `LoadState`
>    --> /md0/code/reth/crates/rpc/rpc-eth-api/src/helpers/state.rs:227:5
>     |
> 227 | /     fn transaction_count(
> 228 | |         &self,
> 229 | |         address: Address,
> 230 | |         block_id: Option<BlockId>,
> 231 | |     ) -> impl Future<Output = E...
> 232 | |     where
> 233 | |         Self: SpawnBlocking,
>     | |____________________________^
> help: disambiguate the method for candidate #1
>     |
> 227 |             let mid_nonce = EthApiServer::transaction_count(&self.eth, sender, Some(BlockId::Number(BlockNumberOrTag::Number(mid))))
>     |                             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> help: disambiguate the method for candidate #2
>     |
> 227 |             let mid_nonce = LoadState::transaction_count(&self.eth, sender, Some(BlockId::Number(BlockNumberOrTag::Number(mid))))
> ```
> 
>  > move the binary search algorithm out to a new method
> > tests for the binary search method
> 
> I'll try to move out and add tests.
> 
> > before checking mid block, check high block and return early if the high block has nonce zero
> 
> not fully understand, we have already check the highest block before the binary search.
> 
> > preferably some heuristics for optimisation: if high block has tx nonce 1 - 5, start with mid set to ¾ high.
> 
> in this pr, I'll try to keep it is as simple, and maybe optimize the binary search later(with some data analyze)

ah, ok, thanks for the explanation, then I get why dismissed

File: crates/rpc/rpc/src/otterscan.rs (hunk 6 above)
Comment (jsvisa) [reply crates/rpc/rpc/src/otterscan.rs:247]: Erigon is searching nonce from the `AccountsHistory` table by tx index, but we are searching in the block body instead, so the implementation is different. 

eg: if we found the nonce changed in block 10, then iterate the txs in block 9 can't get anything.

Comment (emhane) [reply crates/rpc/rpc/src/otterscan.rs:251]: if you added some test to show this, I would be strongly convinced! you can compare without output from region tests for same inputs

File: crates/rpc/rpc/src/otterscan.rs (hunk 4 above)
Comment (emhane) [reply crates/rpc/rpc/src/otterscan.rs:242]: there is no reason to keep looping when we have reached the desired condition

Comment (jsvisa) [reply crates/rpc/rpc/src/otterscan.rs:242]: Sorry, haven't changed it, I want the binary search to be more common, so the getContractCreator can reuse it. But before that I'll first add some tests

Comment (jsvisa) [reply crates/rpc/rpc/src/otterscan.rs:242]: > there is no reason to keep looping when we have reached the desired condition

I see what the problem is, if the nonce was changed in block 6, 10, so we get the same result in block 6, 7, 8, 9. so I think the condition `mid_nonce == nonce + 1` isn't match what we need(the mid can be any of 6..=9).

**Comment by jsvisa** [comment]
@emhane I've adjusted the code and added a simple unit test, but the test case seems too simple, I want to add some integration tests, eg: start a local chain, and then send some transaction, then client request the rpc to fetch the txhash. I don't know where to start with 
it. Do you have any references to learn?


btw, I added a simple python script to test the rpc with my local holesky node, seems everything is ok now.

```py
#!/usr/bin/env python

import requests
import argparse
import logging
from web3 import Web3

logging.basicConfig(format="[%(asctime)s] - %(levelname)s - %(message)s", level=logging.INFO)


def generate_rpc(tx):
    return {
        "jsonrpc": "2.0",
        "id": "0x" + tx["hash"].hex(),
        "method": "ots_getTransactionBySenderAndNonce",
        "params": [tx["from"], tx["nonce"]],
    }


def batch_get_txBySenderAndNonce(provider, calls):
    return requests.post(provider, json=calls).json()


def main():
    parser = argparse.ArgumentParser(
        formatter_class=argparse.ArgumentDefaultsHelpFormatter
    )
    parser.add_argument(
        "--provider",
        "-p",
        default="http://127.0.0.1:8545",
        help="HTTPProvider URL",
    )
    parser.add_argument(
        "--start-block",
        type=int,
        default=1,
        help="Start block number",
    )
    parser.add_argument(
        "--end-block",
        type=int,
        help="End block number",
    )
    args = parser.parse_args()

    w3 = Web3(Web3.HTTPProvider(args.provider))

    end_block = args.end_block or w3.eth.block_number

    for block in range(args.start_block, end_block):
        if block % 100 == 0:
            logging.info(f"check block: {block} ...")
        txs = w3.eth.get_block(block, True)["transactions"]  # type: ignore
        if len(txs) == 0:
            continue

        calls = [generate_rpc(tx) for tx in txs]  # type: ignore
        results = batch_get_txBySenderAndNonce(args.provider, calls)
        diff = [e for e in results if e["id"] != e["result"]]
        if len(diff) == 0:
            continue

        logging.error(f"check block {block} failed: #{len(diff)}/#{len(results)}")
        for d in diff:
            logging.info(d)


if __name__ == "__main__":
    main()
```

**Review by mattsse — APPROVED** [review]
ty!

this should work

**Pull request merged**